File: src/kernel.c
/*
 * MAGIos Kernel - Swift Integration Version
 * This is the hybrid C/Swift kernel implementation
 * It provides basic bootstrapping in C and then calls Swift kernel functions
 */

#include <stddef.h> /* For size_t and NULL definitions */
#include <stdint.h> /* For fixed-width integer types (uint8_t, uint16_t, etc.) */

/* Include our Swift bridge header for interoperability */
#include "include/kernel_bridge.h"

/* === MULTIBOOT INFORMATION HANDLING ===
 * CRITICAL: Process multiboot information from GRUB
 * This data is passed from the bootloader and contains system information
 */
typedef struct multiboot_info {
  uint32_t flags;
  uint32_t mem_lower;
  uint32_t mem_upper;
  uint32_t boot_device;
  uint32_t cmdline;
  uint32_t mods_count;
  uint32_t mods_addr;
  /* ... other fields as needed ... */
} multiboot_info_t;

/* === LEGACY C FUNCTIONS ===
 * These remain in C for compatibility and low-level operations
 * Some may be gradually moved to Swift in future phases
 */

/* Legacy VGA helper functions for fallback/debugging */
static inline uint8_t vga_entry_color_c(vga_color_t fg, vga_color_t bg) {
  return fg | bg << 4;
}

static inline uint16_t vga_entry_c(unsigned char uc, uint8_t color) {
  return (uint16_t)uc | (uint16_t)color << 8;
}

/* === EMERGENCY FALLBACK FUNCTIONS ===
 * These provide basic output if Swift kernel fails
 * CRITICAL: These ensure we can always display error messages
 */
static void emergency_print(const char *message) {
  volatile uint16_t *vga_buffer = (uint16_t *)VGA_MEMORY;
  uint8_t color = VGA_ENTRY_COLOR(VGA_COLOR_LIGHT_RED, VGA_COLOR_BLACK);

  /* Clear first line for emergency message */
  for (int i = 0; i < VGA_WIDTH; i++) {
    vga_buffer[i] = VGA_ENTRY(' ', color);
  }

  /* Display emergency message */
  int pos = 0;
  while (message[pos] && pos < VGA_WIDTH - 1) {
    vga_buffer[pos] = VGA_ENTRY(message[pos], color);
    pos++;
  }
}

/* === KERNEL PANIC FUNCTION ===
 * CRITICAL: Handle kernel panic situations
 */
static void kernel_panic(const char *message) {
  /* Disable interrupts to prevent further issues */
  __asm__ volatile("cli");

  emergency_print("KERNEL PANIC: ");
  emergency_print(message);

  /* Halt the system */
  while (1) {
    __asm__ volatile("hlt");
  }
}

/* === SWIFT KERNEL INITIALIZATION ===
 * CRITICAL: Initialize and call Swift kernel components
 */
static void initialize_swift_kernel(void) {
  /* Attempt to initialize Swift terminal system */
  __asm__ volatile("" ::: "memory"); /* Memory barrier */

  /* Call Swift kernel main function */
  swift_kernel_main();

  __asm__ volatile("" ::: "memory"); /* Memory barrier */
}

/* === BOOT INFORMATION PROCESSING ===
 * Extract and prepare boot information for Swift kernel
 */
static boot_info_t process_multiboot_info(uint32_t magic,
                                          multiboot_info_t *mbi) {
  boot_info_t boot_info = {0};

  boot_info.magic = magic;

  if (mbi && (magic == 0x2BADB002)) {
    boot_info.flags = mbi->flags;
    if (mbi->flags & 0x01) {
      boot_info.memory_lower = mbi->mem_lower;
      boot_info.memory_upper = mbi->mem_upper;
    }
  }

  return boot_info;
}

/* === MAIN KERNEL FUNCTION ===
 * CRITICAL: This is called from our assembly boot code
 * This function coordinates between C bootstrap and Swift kernel
 */
void kernel_main(void) {
  /* === EARLY INITIALIZATION === */
  /* Set up basic error handling */

  /* === SWIFT KERNEL INTEGRATION === */
  /* Try to initialize and run Swift kernel */
  __asm__ volatile("" ::: "memory"); /* Compiler barrier */

  /* Initialize Swift kernel - this does the main work */
  initialize_swift_kernel();

  /* === POST-SWIFT PROCESSING === */
  /* Swift kernel has completed its initialization and display */
  /* Now we handle the final system state */

  /* Display C kernel status message */
  swift_terminal_setcolor(
      VGA_ENTRY_COLOR(VGA_COLOR_DARK_GREY, VGA_COLOR_BLACK));
  swift_terminal_writestring("C kernel: Swift integration successful\n");

  /* === INFINITE IDLE LOOP ===
   * CRITICAL: The kernel must never return from kernel_main
   * We halt the CPU to save power while waiting for interrupts
   */
  __asm__ volatile("cli"); /* Disable interrupts for clean halt */

  while (1) {
    __asm__ volatile(
        "hlt"); /* Halt instruction - stops CPU until next interrupt */
                /*
                 * NOTE: Since we have interrupts disabled (cli was called),
                 * only Non-Maskable Interrupts (NMI) can wake us up.
                 * This is fine for our current "Hello World" OS.
                 * Future versions will enable interrupts and implement proper idle
                 * handling.
                 */
  }

  /* This point should never be reached */
}

/* === ALTERNATIVE ENTRY POINTS ===
 * These provide different kernel modes for testing and development
 */

/* C-only kernel mode (for fallback testing) */
void kernel_main_c_only(void) {
  /* Initialize VGA manually for C-only mode */
  volatile uint16_t *terminal_buffer = (uint16_t *)VGA_MEMORY;
  uint8_t color = VGA_ENTRY_COLOR(VGA_COLOR_LIGHT_CYAN, VGA_COLOR_BLACK);

  /* Clear screen */
  for (size_t y = 0; y < VGA_HEIGHT; y++) {
    for (size_t x = 0; x < VGA_WIDTH; x++) {
      const size_t index = y * VGA_WIDTH + x;
      terminal_buffer[index] = VGA_ENTRY(' ', color);
    }
  }

  /* Display C-only message */
  const char *message = "MAGIos - C-only mode (Swift disabled)";
  size_t row = 0;
  size_t col = 0;

  for (size_t i = 0; message[i] != '\0'; i++) {
    if (message[i] == '\n') {
      row++;
      col = 0;
      continue;
    }

    const size_t index = row * VGA_WIDTH + col;
    terminal_buffer[index] = VGA_ENTRY(message[i], color);
    col++;
  }

  /* Infinite loop */
  while (1) {
    __asm__ volatile("hlt");
  }
}

/* === DEBUGGING FUNCTIONS ===
 * These are available for development and troubleshooting
 */

#ifdef DEBUG
void debug_print(const char *message) {
  swift_terminal_setcolor(VGA_ENTRY_COLOR(VGA_COLOR_BROWN, VGA_COLOR_BLACK));
  swift_terminal_writestring("[DEBUG] ");
  swift_terminal_writestring(message);
  swift_terminal_writestring("\n");
}
#endif

/* === SYSTEM INFORMATION FUNCTIONS ===
 * These provide system status and diagnostic information
 */
void display_system_diagnostics(void) {
  swift_terminal_setcolor(
      VGA_ENTRY_COLOR(VGA_COLOR_LIGHT_BROWN, VGA_COLOR_BLACK));
  swift_terminal_writestring("System Diagnostics:\n");

  swift_terminal_setcolor(VGA_ENTRY_COLOR(VGA_COLOR_WHITE, VGA_COLOR_BLACK));
  swift_terminal_writestring("- C Kernel: Active (integration mode)\n");
  swift_terminal_writestring("- Swift Kernel: Active (embedded mode)\n");
  swift_terminal_writestring("- Boot Protocol: Multiboot v1\n");
  swift_terminal_writestring("- Architecture: i686-elf\n\n");
}

/*
 * === SUMMARY OF HYBRID KERNEL ARCHITECTURE ===
 *
 * CRITICAL C COMPONENTS:
 * 1. Multiboot header processing (boot.s)
 * 2. Early initialization and error handling
 * 3. Swift kernel initialization and coordination
 * 4. Final system halt and infinite loop
 * 5. Emergency fallback functions
 *
 * SWIFT COMPONENTS:
 * 1. Main terminal and VGA handling
 * 2. MAGI system display and theming
 * 3. Memory-safe operations
 * 4. Modern language features and type safety
 *
 * INTEGRATION BENEFITS:
 * 1. C provides low-level system control
 * 2. Swift provides safe, high-level operations
 * 3. Gradual migration path from C to Swift
 * 4. Emergency fallback to C-only mode
 * 5. Best of both worlds approach
 *
 * CALLING SEQUENCE:
 * boot.s -> kernel_main() -> initialize_swift_kernel() -> swift_kernel_main()
 *        -> [Swift does main work] -> return to C -> infinite halt loop
 */

File: src/Kernel.swift
/*
 * MAGIos Swift Kernel - Production Embedded Swift Implementation
 * Evangelion-themed operating system kernel with Embedded Swift
 * Provides VGA terminal management and MAGI system interface
 */

// MARK: - VGA Hardware Constants

/// VGA text mode hardware specifications
private let VGA_WIDTH: Int = 80
private let VGA_HEIGHT: Int = 25
private let VGA_MEMORY: UInt32 = 0xB8000

// MARK: - VGA Hardware Color System

/// VGA 4-bit color palette (hardware specification)
@frozen
enum VGAColor: UInt8 {
    case black = 0
    case blue = 1
    case green = 2
    case cyan = 3
    case red = 4
    case magenta = 5
    case brown = 6
    case lightGrey = 7
    case darkGrey = 8
    case lightBlue = 9
    case lightGreen = 10
    case lightCyan = 11
    case lightRed = 12
    case lightMagenta = 13
    case lightBrown = 14
    case white = 15
}

// MARK: - VGA Hardware Interface

/// Combine foreground and background colors (BBBBFFFF format)
@inline(__always)
private func vgaEntryColor(foreground: VGAColor, background: VGAColor) -> UInt8 {
    return foreground.rawValue | (background.rawValue << 4)
}

/// Create VGA character entry (2 bytes: character + attributes)
@inline(__always)
private func vgaEntry(character: UInt8, color: UInt8) -> UInt16 {
    return UInt16(character) | (UInt16(color) << 8)
}

// MARK: - Terminal Hardware State

/// VGA terminal state management
private final class TerminalState {
    var row: Int = 0
    var column: Int = 0
    var color: UInt8 = 0
    var buffer: UnsafeMutablePointer<UInt16>

    init() {
        self.buffer = UnsafeMutablePointer<UInt16>(bitPattern: Int(VGA_MEMORY))!
        self.color = vgaEntryColor(foreground: .lightCyan, background: .black)
    }
}

/// Global terminal hardware interface
private let terminal = TerminalState()

// MARK: - Terminal Hardware Operations

/// Initialize VGA terminal hardware
private func terminalInitialize() {
    terminal.row = 0
    terminal.column = 0
    terminal.color = vgaEntryColor(foreground: .lightCyan, background: .black)

    // Clear VGA buffer
    for y in 0..<VGA_HEIGHT {
        for x in 0..<VGA_WIDTH {
            let index = y * VGA_WIDTH + x
            terminal.buffer[index] = vgaEntry(character: UInt8(ascii: " "), color: terminal.color)
        }
    }
}

/// Write character to VGA buffer at cursor position
private func terminalPutchar(_ character: UInt8) {
    if character == UInt8(ascii: "\n") {
        terminal.column = 0
        terminal.row += 1
        if terminal.row >= VGA_HEIGHT {
            terminal.row = 0
        }
        return
    }

    let index = terminal.row * VGA_WIDTH + terminal.column
    terminal.buffer[index] = vgaEntry(character: character, color: terminal.color)

    terminal.column += 1
    if terminal.column >= VGA_WIDTH {
        terminal.column = 0
        terminal.row += 1
        if terminal.row >= VGA_HEIGHT {
            terminal.row = 0
        }
    }
}

/// Write buffer to terminal
private func terminalWrite(_ data: UnsafePointer<UInt8>, size: Int) {
    for i in 0..<size {
        terminalPutchar(data[i])
    }
}

/// Calculate C string length
private func swiftStrlen(_ str: UnsafePointer<UInt8>) -> Int {
    var length = 0
    while str[length] != 0 {
        length += 1
    }
    return length
}

/// Write null-terminated string with formatting
private func terminalWritestring(_ data: UnsafePointer<UInt8>) {
    let dataLength = swiftStrlen(data)

    // Add spacing for alignment
    for _ in 0..<5 {
        terminalPutchar(UInt8(ascii: " "))
    }

    for i in 0..<dataLength {
        terminalPutchar(data[i])
    }
}

/// Set terminal color attributes
private func terminalSetcolor(_ color: UInt8) {
    terminal.color = color
}

// MARK: - MAGI System Interface

/// MAGI system initialization sequence
private func displayMAGIStartup() {
    terminalSetcolor(vgaEntryColor(foreground: .lightRed, background: .black))
    terminalWritestring("======================================\n")
    terminalWritestring("MAGI SYSTEM STARTUP SEQUENCE INITIATED\n")
    terminalWritestring("======================================\n\n")

    terminalSetcolor(vgaEntryColor(foreground: .lightCyan, background: .black))
    terminalWritestring("CASPER... ")
    terminalSetcolor(vgaEntryColor(foreground: .lightGreen, background: .black))
    terminalWritestring("ONLINE\n")

    terminalSetcolor(vgaEntryColor(foreground: .lightCyan, background: .black))
    terminalWritestring("MELCHIOR... ")
    terminalSetcolor(vgaEntryColor(foreground: .lightGreen, background: .black))
    terminalWritestring("ONLINE\n")

    terminalSetcolor(vgaEntryColor(foreground: .lightCyan, background: .black))
    terminalWritestring("BALTHASAR... ")
    terminalSetcolor(vgaEntryColor(foreground: .lightGreen, background: .black))
    terminalWritestring("ONLINE\n\n")
}

/// Display kernel status information
private func displaySystemInfo() {
    terminalSetcolor(vgaEntryColor(foreground: .lightMagenta, background: .black))
    terminalWritestring("MAGIos v0.0.1 - Swift Edition\n")
    terminalSetcolor(vgaEntryColor(foreground: .white, background: .black))
    terminalWritestring("Boot Successful (Swift Kernel Active)\n\n")

    terminalSetcolor(vgaEntryColor(foreground: .lightBrown, background: .black))
    terminalWritestring("System Status:\n")
    terminalSetcolor(vgaEntryColor(foreground: .white, background: .black))
    terminalWritestring("- Kernel: Swift Embedded + C Assembly\n")
    terminalWritestring("- Memory: 1MB+ physical (Swift managed)\n")
    terminalWritestring("- VGA: 80x25 text mode (Swift controlled)\n")
    terminalWritestring("- Runtime: Embedded Swift (ARC disabled)\n\n")
}

/// Display final initialization status
private func displayFinalMessage() {
    terminalSetcolor(vgaEntryColor(foreground: .lightRed, background: .black))
    terminalWritestring("Hello, World from Swift MAGIos!\n")
    terminalSetcolor(vgaEntryColor(foreground: .lightGrey, background: .black))
    terminalWritestring("Swift kernel initialized successfully...\n")

    terminalSetcolor(vgaEntryColor(foreground: .lightMagenta, background: .black))
    terminalWritestring("AT Field operational. Pattern Blue.\n")
    terminalSetcolor(vgaEntryColor(foreground: .lightGrey, background: .black))
    terminalWritestring("System entering infinite idle loop...\n")
}

// MARK: - Kernel Entry Points

/// Main Swift kernel entry point (called from C bootstrap)
@_cdecl("swift_kernel_main")
public func swiftKernelMain() {
    terminalInitialize()
    displayMAGIStartup()
    displaySystemInfo()
    displayFinalMessage()
}

/// C-compatible terminal string output
@_cdecl("swift_terminal_writestring")
public func swiftTerminalWritestring(_ data: UnsafePointer<UInt8>) {
    terminalWritestring(data)
}

/// C-compatible terminal color control
@_cdecl("swift_terminal_setcolor")
public func swiftTerminalSetcolor(_ color: UInt8) {
    terminalSetcolor(color)
}

/// C-compatible terminal initialization
@_cdecl("swift_terminal_initialize")
public func swiftTerminalInitialize() {
    terminalInitialize()
}

/*
 * MAGIos Swift Kernel Implementation Summary
 *
 * Core Features:
 * - Hardware-direct VGA text mode control
 * - MAGI system initialization sequence
 * - C/Swift interoperability layer
 * - Memory-safe pointer operations
 * - Embedded Swift optimizations
 *
 * Architecture:
 * - Freestanding embedded Swift environment
 * - Direct hardware memory access (0xB8000)
 * - Minimal runtime footprint
 * - Compatible with C bootstrap code
 */
